File: cssharp/Services/MessageProcessor.cs
using System;
using System.Globalization;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace NotifyMessages;

/// Сервис обработки и локализации сообщений
/// Отвечает за:
/// - подстановку языковых шаблонов из Config.LanguageMessages
/// - замену тегов {MAP}/{TIME}/{DATE}/{SERVERNAME}/{IP}/{PORT}/{MAXPLAYERS}/{PLAYERS}
/// - замену \n на спец-символ для отображения в центре
/// - замену имён карт согласно Config.MapsName
///
/// ВАЖНО: вызывать только из главного потока — ReplaceMessageTags читает IServerInfoSource,
/// а за ним стоят нативы движка (имя карты, ConVar, список игроков).
///
/// Фреймворка здесь нет ни одной строкой: всё, что зависит от движка, приходит через
/// IServerInfoSource. Поэтому класс совпадает в cssharp/ и swiftly/ и проверяется
/// тестами на любой машине.
public sealed class MessageProcessor
{
    // internal, а не private: диагностика шаблонов обязана видеть ровно те же теги,
    // что и подстановка, иначе анализатор начнёт врать.
    internal static readonly Regex TagPattern = new Regex(@"\{([^}]*)\}", RegexOptions.Compiled);

    /// Теги, которые ReplaceMessageTags подставляет в любом сообщении, без всякого контекста.
    /// Список — источник истины для диагностики шаблонов, держать синхронно с ReplaceMessageTags.
    internal static readonly string[] SystemTags =
    {
        "{MAP}", "{TIME}", "{DATE}", "{SERVERNAME}", "{IP}", "{PORT}", "{MAXPLAYERS}", "{PLAYERS}"
    };

    private static readonly HashSet<string> SystemTagSet =
        new(SystemTags, StringComparer.OrdinalIgnoreCase);

    internal static bool IsSystemTag(string tag) => SystemTagSet.Contains(tag);

    private readonly Config _config;
    private readonly Func<ulong, string?> _getIsoCodeBySteamId;
    private readonly IServerInfoSource _server;

    public MessageProcessor(Config config, Func<ulong, string?> getIsoCodeBySteamId, IServerInfoSource server)
    {
        _config = config;
        _getIsoCodeBySteamId = getIsoCodeBySteamId;
        _server = server;
    }

    /// Получить ISO-код для игрока (для кеширования)
    public string? GetIsoCodeBySteamId(ulong steamId)
    {
        return _getIsoCodeBySteamId(steamId);
    }

    /// Применяет локализацию, контекстные значения, системные теги и рендер под канал.
    ///
    /// Порядок частей не произволен:
    ///   язык -> контекстные значения -> системные теги -> рендер.
    /// Значения подставляются ДО рендера, потому что сами содержат теги ({RED}Terrorists{DEFAULT}).
    /// Раньше их подставляли после ProcessMessage, и игрок видел эти теги текстом.
    public string ProcessMessage(string message, ulong steamId,
        MessageType channel = MessageType.Chat,
        IReadOnlyDictionary<string, string>? values = null)
    {
        if (string.IsNullOrEmpty(message))
            return string.Empty;

        var result = ApplyLanguage(message, steamId);
        result = ApplyValues(result, values, channel);
        result = ReplaceMessageTags(result);

        return Render(result, channel);
    }

    private string ApplyLanguage(string message, ulong steamId)
    {
        if (_config.LanguageMessages == null) return message;

        var matches = TagPattern.Matches(message);
        foreach (Match match in matches)
        {
            var tag = match.Groups[0].Value;
            var tagName = match.Groups[1].Value;

            if (!_config.LanguageMessages.TryGetValue(tagName, out var language))
                continue;

            var isoCode = steamId > 0 ? _getIsoCodeBySteamId(steamId) : _config.DefaultLang;

            if (isoCode != null && language.TryGetValue(isoCode, out var replacement))
                message = message.Replace(tag, replacement, StringComparison.Ordinal);
            else if (_config.DefaultLang != null && language.TryGetValue(_config.DefaultLang, out var defReplacement))
                message = message.Replace(tag, defReplacement, StringComparison.Ordinal);
        }

        return message;
    }

    /// Подстановка контекстных значений ({PLAYERNAME}, {TEAM}, {SECONDS}, ...).
    /// Регистр игнорируется: дефолтный Messages.json пишет часть тегов строчными.
    /// Для HTML-панели значения экранируются — ник игрока это недоверенные данные.
    internal static string ApplyValues(string message, IReadOnlyDictionary<string, string>? values,
        MessageType channel)
    {
        if (values == null || values.Count == 0) return message;

        foreach (var (tag, value) in values)
        {
            var safe = channel == MessageType.CenterHtml ? TextFormatter.EscapeHtml(value) : value;
            message = TextFormatter.ReplaceIgnoreCase(message, tag, safe);
        }

        return message;
    }

    /// Финальный рендер. У каналов разные грамматики, одна строка не может быть верной во всех:
    /// чат понимает управляющие байты, HTML-панель — разметку, центр и консоль — только текст.
    internal static string Render(string text, MessageType channel) => channel switch
    {
        MessageType.Chat => text.ReplaceColorTags().Replace("\n", "\u2029", StringComparison.Ordinal),
        MessageType.CenterHtml => TextFormatter.ToCenterHtml(text),
        // В консоли перенос строки — настоящий \n, а не U+2029
        MessageType.Console => TextFormatter.RemoveColorTags(text),
        _ => TextFormatter.RemoveColorTags(text).Replace("\n", "\u2029", StringComparison.Ordinal)
    };

    /// Возвращает случайный ШАБЛОН из набора (Join/Leave) на языке получателя.
    /// Значения ({PLAYERNAME}, {COUNTRY}, {CITY}) подставляет ProcessMessage — единственная
    /// точка подстановки на весь плагин.
    public string GetRandomLocalizedMessage(Dictionary<string, List<string>>? messages, ulong recipientSteamId)
    {
        if (messages == null || messages.Count == 0) return string.Empty;

        var lang = _config.DefaultLang ?? "US";
        var iso = _getIsoCodeBySteamId(recipientSteamId);
        if (iso != null && messages.ContainsKey(iso))
            lang = iso;

        if (!messages.TryGetValue(lang, out var messageList) || messageList.Count == 0) return string.Empty;

        return messageList[Random.Shared.Next(messageList.Count)];
    }

    /// Заменяет системные теги и имена карт.
    /// Каждый плейсхолдер резолвится ТОЛЬКО если реально встречается в строке —
    /// иначе на каждое сообщение уходило по 3 ConVar.Find + GetPlayers() + GetMapName().
    public string ReplaceMessageTags(string message)
    {
        if (string.IsNullOrEmpty(message)) return message;

        var result = message;

        if (result.Contains("{MAP}", StringComparison.Ordinal))
            result = result.Replace("{MAP}", _server.MapName, StringComparison.Ordinal);

        if (result.Contains("{TIME}", StringComparison.Ordinal))
            result = result.Replace("{TIME}", DateTime.Now.ToString("HH:mm:ss", CultureInfo.InvariantCulture));

        if (result.Contains("{DATE}", StringComparison.Ordinal))
            result = result.Replace("{DATE}", DateTime.Now.ToString("dd.MM.yyyy", CultureInfo.InvariantCulture));

        if (result.Contains("{SERVERNAME}", StringComparison.Ordinal))
            result = result.Replace("{SERVERNAME}", _server.Hostname, StringComparison.Ordinal);

        if (result.Contains("{IP}", StringComparison.Ordinal))
            result = result.Replace("{IP}", _server.Ip, StringComparison.Ordinal);

        if (result.Contains("{PORT}", StringComparison.Ordinal))
            result = result.Replace("{PORT}", _server.Port, StringComparison.Ordinal);

        if (result.Contains("{MAXPLAYERS}", StringComparison.Ordinal))
            result = result.Replace("{MAXPLAYERS}", _server.MaxPlayers.ToString(CultureInfo.InvariantCulture),
                StringComparison.Ordinal);

        if (result.Contains("{PLAYERS}", StringComparison.Ordinal))
            result = result.Replace("{PLAYERS}", _server.Players.ToString(CultureInfo.InvariantCulture),
                StringComparison.Ordinal);

        if (_config.MapsName != null)
        {
            foreach (var (key, niceName) in _config.MapsName)
            {
                // Regex дорогой — не запускаем его для карт, которых нет в строке.
                // Замена через MatchEvaluator, а не строкой: строка замены — это шаблон,
                // в котором "$1" и "$$" имеют смысл, и красивое имя карты с долларом
                // превращалось бы в мусор.
                if (result.Contains(key, StringComparison.Ordinal))
                    result = Regex.Replace(result, $@"\b{Regex.Escape(key)}\b", _ => niceName);
            }
        }

        return result;
    }
}

File: metamod/src/core/message_processor.cpp
#include "core/message_processor.h"

#include "core/text_formatter.h"

#include <cctype>
#include <cstdint>
#include <ctime>
#include <random>

namespace nm {
namespace {

constexpr const char* kSystemTags[] = {"{MAP}",  "{TIME}", "{DATE}",       "{SERVERNAME}",
                                       "{IP}",   "{PORT}", "{MAXPLAYERS}", "{PLAYERS}"};

constexpr const char* kParagraph = "\xE2\x80\xA9";  // U+2029: перенос строки чата и центра

bool EqualsIgnoreCase(const std::string& a, const char* b) {
    size_t i = 0;
    for (; i < a.size() && b[i] != '\0'; ++i) {
        if (std::tolower(static_cast<unsigned char>(a[i])) !=
            std::tolower(static_cast<unsigned char>(b[i]))) {
            return false;
        }
    }
    return i == a.size() && b[i] == '\0';
}

// Кодовая точка UTF-8, начинающаяся в позиции at (или заканчивающаяся перед ней).
// Битая последовательность читается как один байт — для границы слова этого хватает.
uint32_t DecodeAt(const std::string& s, size_t at) {
    const unsigned char b0 = static_cast<unsigned char>(s[at]);
    if (b0 < 0x80) return b0;
    const int extra = b0 >= 0xF0 ? 3 : b0 >= 0xE0 ? 2 : b0 >= 0xC0 ? 1 : 0;
    if (extra == 0 || at + extra >= s.size()) return b0;
    uint32_t cp = extra == 3 ? (b0 & 0x07) : extra == 2 ? (b0 & 0x0F) : (b0 & 0x1F);
    for (int i = 1; i <= extra; ++i) {
        const unsigned char b = static_cast<unsigned char>(s[at + i]);
        if ((b & 0xC0) != 0x80) return b0;
        cp = (cp << 6) | (b & 0x3F);
    }
    return cp;
}

uint32_t DecodeBefore(const std::string& s, size_t at) {
    size_t start = at - 1;
    while (start > 0 && (static_cast<unsigned char>(s[start]) & 0xC0) == 0x80 && at - start < 4) {
        --start;
    }
    return DecodeAt(s, start);
}

// Символ «слова» для \b в .NET: буквы, цифры и подчёркивание.
// ponytail: для не-ASCII — эвристика по блокам (пунктуация, стрелки, символы,
// эмодзи — не слово; остальное — слово). Полная таблица Unicode ради границы
// имени карты не нужна; расхождение с C# возможно только у экзотических символов.
bool IsWordChar(uint32_t cp) {
    if (cp < 0x80) return std::isalnum(static_cast<int>(cp)) || cp == '_';
    if (cp <= 0xBF || cp == 0xD7 || cp == 0xF7) return false;       // Latin-1: знаки
    if (cp >= 0x2000 && cp <= 0x2BFF) return false;                  // пунктуация, стрелки, символы
    if (cp >= 0x2E00 && cp <= 0x2E7F) return false;
    if (cp >= 0x3000 && cp <= 0x303F) return false;                  // CJK-пунктуация
    if (cp >= 0xFE30 && cp <= 0xFE4F) return false;
    if (cp >= 0xFF00 && cp <= 0xFF0F) return false;
    if (cp >= 0x1F000 && cp <= 0x1FAFF) return false;                // эмодзи
    return true;
}

// Regex.Replace(text, $@"\b{key}\b", _ => niceName). Замена функцией, а не строкой:
// в строке замены "$" — спецсимвол, и имя карты с долларом превращалось в мусор.
std::string ReplaceWholeWord(const std::string& text, const std::string& key,
                             const std::string& replacement) {
    if (key.empty()) return text;

    std::string out;
    size_t last = 0;
    size_t at = text.find(key);
    while (at != std::string::npos) {
        const size_t end = at + key.size();
        // \b стоит между символом слова и не-словом. Для ключа, начинающегося
        // не с символа слова ("-map"), граница там, где слева слово.
        const bool keyStartsWord = IsWordChar(DecodeAt(key, 0));
        const bool keyEndsWord = IsWordChar(DecodeBefore(key, key.size()));
        const bool leftWord = at > 0 && IsWordChar(DecodeBefore(text, at));
        const bool rightWord = end < text.size() && IsWordChar(DecodeAt(text, end));

        if (leftWord != keyStartsWord && rightWord != keyEndsWord) {
            out.append(text, last, at - last);
            out += replacement;
            last = end;
            at = text.find(key, end);
        } else {
            at = text.find(key, at + 1);
        }
    }
    if (last == 0 && out.empty()) return text;
    out.append(text, last, std::string::npos);
    return out;
}

// Локальное время сервера: {TIME} и {DATE} в C# идут от DateTime.Now
std::string FormatNow(const char* format) {
    const std::time_t now = std::time(nullptr);
    std::tm local{};
#ifdef _WIN32
    localtime_s(&local, &now);
#else
    localtime_r(&now, &local);
#endif
    char buffer[32];
    const size_t size = std::strftime(buffer, sizeof(buffer), format, &local);
    return std::string(buffer, size);
}

std::mt19937& Random() {
    // Один генератор на процесс; зовётся только из главного потока
    static std::mt19937 generator{std::random_device{}()};
    return generator;
}

}  // namespace

bool MessageProcessor::IsSystemTag(const std::string& tag) {
    for (const char* system : kSystemTags) {
        if (EqualsIgnoreCase(tag, system)) return true;
    }
    return false;
}

std::vector<TagMatch> MessageProcessor::FindTags(const std::string& text) {
    std::vector<TagMatch> matches;
    size_t from = 0;
    while (true) {
        const size_t open = text.find('{', from);
        if (open == std::string::npos) break;
        const size_t close = text.find('}', open + 1);
        if (close == std::string::npos) break;
        matches.push_back({text.substr(open, close - open + 1), text.substr(open + 1, close - open - 1)});
        from = close + 1;
    }
    return matches;
}

std::string MessageProcessor::ProcessMessage(const std::string& message, uint64_t steamId,
                                             MessageType channel, const Values* values) const {
    if (message.empty()) return std::string();

    std::string result = ApplyLanguage(message, steamId);
    result = ApplyValues(result, values, channel);
    result = ReplaceMessageTags(result);
    return Render(result, channel);
}

std::string MessageProcessor::ApplyLanguage(const std::string& message, uint64_t steamId) const {
    if (_config->languageMessages.empty()) return message;

    std::string result = message;
    for (const TagMatch& match : FindTags(message)) {
        const auto language = _config->languageMessages.find(match.name);
        if (language == _config->languageMessages.end()) continue;

        const std::string iso = steamId > 0 ? _language(steamId) : _config->defaultLang;

        const auto exact = language->second.find(iso);
        if (exact != language->second.end()) {
            result = text::ReplaceOrdinal(result, match.tag, exact->second);
            continue;
        }

        const auto fallback = language->second.find(_config->defaultLang);
        if (fallback != language->second.end()) {
            result = text::ReplaceOrdinal(result, match.tag, fallback->second);
        }
    }
    return result;
}

std::string MessageProcessor::ApplyValues(const std::string& message, const Values* values,
                                          MessageType channel) {
    if (values == nullptr || values->empty()) return message;

    // Регистр игнорируется: дефолтный Messages.json пишет часть тегов строчными
    std::string result = message;
    for (const auto& value : *values) {
        const std::string safe =
            channel == MessageType::CenterHtml ? text::EscapeHtml(value.second) : value.second;
        result = text::ReplaceIgnoreCase(result, value.first, safe);
    }
    return result;
}

std::string MessageProcessor::Render(const std::string& input, MessageType channel) {
    switch (channel) {
        case MessageType::Chat:
            return text::ReplaceOrdinal(text::ReplaceColorTags(input), "\n", kParagraph);
        case MessageType::CenterHtml:
            return text::ToCenterHtml(input);
        case MessageType::Console:
            // В консоли перенос строки — настоящий \n, а не U+2029
            return text::RemoveColorTags(input);
        default:
            return text::ReplaceOrdinal(text::RemoveColorTags(input), "\n", kParagraph);
    }
}

std::string MessageProcessor::GetRandomLocalizedMessage(
    const CiMap<std::vector<std::string>>& messages, uint64_t recipientSteamId) const {
    if (messages.empty()) return std::string();

    std::string lang = _config->defaultLang;
    const std::string iso = _language(recipientSteamId);
    if (!iso.empty() && messages.count(iso) != 0) lang = iso;

    const auto list = messages.find(lang);
    if (list == messages.end() || list->second.empty()) return std::string();

    std::uniform_int_distribution<size_t> pick(0, list->second.size() - 1);
    return list->second[pick(Random())];
}

std::string MessageProcessor::ReplaceMessageTags(const std::string& message) const {
    if (message.empty()) return message;

    std::string result = message;

    if (result.find("{MAP}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{MAP}", _server->MapName());
    }
    if (result.find("{TIME}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{TIME}", FormatNow("%H:%M:%S"));
    }
    if (result.find("{DATE}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{DATE}", FormatNow("%d.%m.%Y"));
    }
    if (result.find("{SERVERNAME}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{SERVERNAME}", _server->Hostname());
    }
    if (result.find("{IP}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{IP}", _server->Ip());
    }
    if (result.find("{PORT}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{PORT}", _server->Port());
    }
    if (result.find("{MAXPLAYERS}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{MAXPLAYERS}", std::to_string(_server->MaxPlayers()));
    }
    if (result.find("{PLAYERS}") != std::string::npos) {
        result = text::ReplaceOrdinal(result, "{PLAYERS}", std::to_string(_server->Players()));
    }

    for (const auto& map : _config->mapsName) {
        if (result.find(map.first) != std::string::npos) {
            result = ReplaceWholeWord(result, map.first, map.second);
        }
    }

    return result;
}

}  // namespace nm
